streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -32,102 +33,198 @@
-            partitionForTopic.put(source.topic(), source);
-        }
-        offsetLimits = new HashMap<>();
-        standbyRestoredOffsets = new ConcurrentHashMap<>();
-        this.isStandby = isStandby;
-        restoreCallbacks = isStandby ? new ConcurrentHashMap<>() : null;
-        recordConverters = isStandby ? new HashMap<>() : null;
-        this.storeToChangelogTopic = new HashMap<>(storeToChangelogTopic);
-
-        baseDir = stateDirectory.directoryForTask(taskId);
-        checkpointFile = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
-        initialLoadedCheckpoints = checkpointFile.read();
-
-        log.trace("Checkpointable offsets read from checkpoint: {}", initialLoadedCheckpoints);
-
-        if (eosEnabled) {
-            // with EOS enabled, there should never be a checkpoint file _during_ processing.
-            // delete the checkpoint file after loading its stored offsets.
-            checkpointFile.delete();
-            checkpointFile = null;
-        }
+        this.baseDir = stateDirectory.directoryForTask(taskId);
+        this.checkpointFile = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
 
         log.debug("Created state store manager for task {}", taskId);
     }
 
+    void registerGlobalStateStores(final List<StateStore> stateStores) {
+        log.debug("Register global stores {}", stateStores);
+        for (final StateStore stateStore : stateStores) {
+            globalStores.put(stateStore.name(), stateStore);
+        }
+    }
+
+    @Override
+    public StateStore getGlobalStore(final String name) {
+        return globalStores.get(name);
+    }
 
-    public static String storeChangelogTopic(final String applicationId,
-                                             final String storeName) {
-        return applicationId + "-" + storeName + STATE_CHANGELOG_TOPIC_SUFFIX;
+    // package-private for test only
+    void initializeStoreOffsetsFromCheckpoint() {
+        try {
+            final Map<TopicPartition, Long> loadedCheckpoints = checkpointFile.read();
+
+            log.trace("Loaded offsets from the checkpoint file: {}", loadedCheckpoints);
+
+            for (final StateStoreMetadata store : stores.values()) {
+                if (store.changelogPartition == null) {
+                    log.info("State store {} is not logged and hence would not be restored", store.stateStore.name());
+                } else {
+                    if (loadedCheckpoints.containsKey(store.changelogPartition)) {
+                        store.setOffset(loadedCheckpoints.remove(store.changelogPartition));
+
+                        log.debug("State store {} initialized from checkpoint with offset {} at changelog {}",
+                            store.stateStore.name(), store.offset, store.changelogPartition);
+                    } else {
+                        // TODO K9113: for EOS when there's no checkpointed offset, we should treat it as TaskCorrupted
```

Comment:
unresolved TODO here. it's unclear to me why we would only consider this task corrupted in EOS mode. It seems like the lack of a checkpoint _file_ just means that we loaded a cached task in an undefined state, and we should discard it and restore. If we have the file, but some of a changelog is missing from it, then maybe it just never got written, before shutdown, or more likely, the topology has changed and the task is corrupted, whether or not we are in EOS.

Reply:
Today we write the checkpoint file before we commit the offsets, so under non-EOS we can just restore from scratch without wiping the local store image since restoring is just overwriting the local store and we can just restore to the end of the changelog (this may result in duplicates but is fine under non-EOS); but with non-EOS we have to first wipe out before restoring from scratch.